tensorflow_quantum/core/src/circuit_parser_qsim_test.cc

```diff
@@ -0,0 +1,480 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "tensorflow_quantum/core/src/circuit_parser_qsim.h"
+
+#include <string>
+
+#include "../qsim/lib/circuit.h"
+#include "../qsim/lib/gates_cirq.h"
+#include "absl/container/flat_hash_map.h"
+#include "absl/strings/numbers.h"
+#include "cirq/google/api/v2/program.pb.h"
+#include "gtest/gtest.h"
+#include "tensorflow/core/lib/core/status.h"
+
+namespace tfq {
+namespace {
+
+typedef absl::flat_hash_map<std::string, std::pair<int, float>> SymbolMap;
+typedef qsim::Cirq::GateCirq<float> QsimGate;
+typedef qsim::Circuit<QsimGate> QsimCircuit;
+
+using ::cirq::google::api::v2::Arg;
+using ::cirq::google::api::v2::Circuit;
+using ::cirq::google::api::v2::Gate;
+using ::cirq::google::api::v2::Moment;
+using ::cirq::google::api::v2::Operation;
+using ::cirq::google::api::v2::Program;
+using ::cirq::google::api::v2::Qubit;
+
+Arg MakeArg(float val) {
+  Arg arg;
+  arg.mutable_arg_value()->set_float_value(val);
+  return arg;
+}
+
+Arg MakeArg(const std::string& val) {
+  Arg arg;
+  arg.set_symbol(val);
+  return arg;
+}
+
+inline void AssertTwoQubitEqual(const QsimGate& a, const QsimGate& b) {
+  for (int i = 0; i < 32; i++) {
+    ASSERT_NEAR(a.matrix[i], b.matrix[i], 1e-5);
+  }
+  ASSERT_EQ(a.qubits[0], b.qubits[0]);
+  ASSERT_EQ(a.qubits[1], b.qubits[1]);
+}
+
+inline void AssertOneQubitEqual(const QsimGate& a, const QsimGate& b) {
+  for (int i = 0; i < 8; i++) {
+    ASSERT_NEAR(a.matrix[i], b.matrix[i], 1e-5);
+  }
+  ASSERT_EQ(a.qubits[0], b.qubits[0]);
+}
+
+class TwoQubitEigenFixture
+    : public ::testing::TestWithParam<std::tuple<
+          std::string, std::function<QsimGate(int, int, int, float, float)>>> {
+};
+
+TEST_P(TwoQubitEigenFixture, TwoEigenGate) {
+  float exp = 1.1234;
+  float gs = 2.2345;
+
+  // Get gate name and reference qsim gate.
+  std::string name = std::get<0>(GetParam());
+  auto ref_gate = std::get<1>(GetParam())(0, 0, 1, exp, gs);
+  Program program_proto;
+  Circuit* circuit_proto = program_proto.mutable_circuit();
+  circuit_proto->set_scheduling_strategy(circuit_proto->MOMENT_BY_MOMENT);
+  Moment* moments_proto = circuit_proto->add_moments();
+
+  // Add gate.
+  Operation* operations_proto = moments_proto->add_operations();
+  Gate* gate_proto = operations_proto->mutable_gate();
+  gate_proto->set_id(name);
+
+  // Set args.
+  google::protobuf::Map<std::string, Arg>* args_proto =
+      operations_proto->mutable_args();
+  (*args_proto)["global_shift"] = MakeArg(gs);
+  (*args_proto)["exponent"] = MakeArg("placeholder");
+  (*args_proto)["exponent_scalar"] = MakeArg(0.5);
+
+  // Set the qubits.
+  Qubit* qubits_proto = operations_proto->add_qubits();
+  qubits_proto->set_id("0");
+  qubits_proto = operations_proto->add_qubits();
+  qubits_proto->set_id("1");
```

Comment:
Could you add a comment describing the circuit diagram?

Reply:
What did you think we should add here could you help me understand ?

Reply:
It was for easy maintenance of this test code. Cirq also drew circuit diagram for their test codes. But I feel it is time-consuming for this now... So let's ignore my thought :) closed.